Fixed cgroup2 path issue in LXC-Top

DevonSchwartz opened the pull request:
We implimented some of the functionality to convert the cgroup paths to cgroup2 paths. However, we do not know how to tell whether system is using cgroup or cgroup2. In addition, some of the paths for cgroup did not have cgroup2 equivalents in the incus directory. 

Fixes #4376
Fixes https://github.com/lxc/lxc/issues/3372

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the allow list.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

stgraber commented:
This pull request has a bunch of changes to unrelated files, you probably want to clean that up.

DevonSchwartz commented:
Just cleaned up the file. I am now going to implement calling cgroup2 paths first and then cgroup afterwards.

DevonSchwartz commented:
Added cgroup2 and cgroup functionality so that warnings are not encountered. If any of the cgroup2 stat calls return an error code, then we call the cgroup container statistics.

stgraber commented:
The PR needs to be rebased as it's incorrectly reverting some changes now.

I also think it'd be best to have each individual metric attempt cgroup2 and fallback to cgroup1 as the current logic you have will fail if you're on a system that's cgroup2 but is somehow missing swap accounting, it will then assume it's cgroup1 and will fail even more as none of those files will be found.

DevonSchwartz commented:
I'll go ahead and push a commit to resolve the conflict in network.c. I must have accidentally reverted to a change before the latest commit to main that fixed a network.c issue.

DevonSchwartz commented:
Just rebased and unstaged irrelevant file changes.

DevonSchwartz commented on src/lxc/tools/lxc_top.c:
```diff
@@ -314,21 +320,58 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
 	ct->stats->kmem_used     = stat_get_int(c, "memory.kmem.usage_in_bytes");
 	ct->stats->kmem_limit    = stat_get_int(c, "memory.kmem.limit_in_bytes");
-	ct->stats->cpu_use_nanos = stat_get_int(c, "cpuacct.usage");
-	ct->stats->cpu_use_user  = stat_match_get_int(c, "cpuacct.stat", "user", 1);
-	ct->stats->cpu_use_sys   = stat_match_get_int(c, "cpuacct.stat", "system", 1);
```
The issue was related to how stats_match_get_int was called in try_cgroup2. I used "user" as the match for both calls. I fixed so that now the function takes a parameter for the type of match.

DevonSchwartz commented:
> Looks reasonable to me. Can you do a rebase to turn this whole thing into a single commit which also has the `Closes #4376` line so it closes the issue?
> 
> @mihalicyn can you review?

Awesome! I squashed into one commit and added the closing line.

ncopa commented:
I have applied this to lxc-6.0.0 in Alpine Linux but it does not seem to solve the issue.
```
# lxc-top
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
Unable to read cgroup item memory.usage_in_bytes
Segmentation fault (core dumped)
```

I have a backtrace:
```
[New LWP 24554]
Core was generated by `lxc-top'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f3667931573 in strlen (s=<optimized out>, s@entry=0x0) at src/string/strlen.c:17

warning: 17     src/string/strlen.c: No such file or directory
(gdb) 
#0  0x00007f3667931573 in strlen (s=<optimized out>, s@entry=0x0) at src/string/strlen.c:17
#1  0x000056275c5f4987 in stat_get_blk_stats (c=c@entry=0x7f366770fd60, item=item@entry=0x56275c5f9a27 "io.stat", stats=0x7f36676d1698, success=success@entry=0x7ffdbd023f1f)
    at ../src/lxc/tools/lxc_top.c:307
#2  0x000056275c5f706f in stats_get (total=0x7ffdbd023f40, ct=0x7f36676d1c30, c=0x7f366770fd60) at ../src/lxc/tools/lxc_top.c:361
#3  lxc_top_main (argc=<optimized out>, argv=<optimized out>) at ../src/lxc/tools/lxc_top.c:663
```

Segfault happens due to `cols[1]` is `NULL` apparently. 
```c
               if (strncmp(cols[1], "Read", strlen(cols[1])) == 0)
                        stats->read += strtoull(cols[2], NULL, 0);
```

DevonSchwartz commented:
> I have applied this to lxc-6.0.0 in Alpine Linux but it does not seem to solve the issue.
> 
> ```
> # lxc-top
> Unable to read cgroup item memory.kmem.usage_in_bytes
> Unable to read cgroup item memory.kmem.limit_in_bytes
> Unable to read cgroup item blkio.throttle.io_serviced
> Unable to read cgroup item memory.kmem.usage_in_bytes
> Unable to read cgroup item memory.kmem.limit_in_bytes
> Unable to read cgroup item blkio.throttle.io_serviced
> Unable to read cgroup item memory.kmem.usage_in_bytes
> Unable to read cgroup item memory.kmem.limit_in_bytes
> Unable to read cgroup item blkio.throttle.io_serviced
> Unable to read cgroup item memory.kmem.usage_in_bytes
> Unable to read cgroup item memory.kmem.limit_in_bytes
> Unable to read cgroup item blkio.throttle.io_serviced
> Unable to read cgroup item memory.kmem.usage_in_bytes
> Unable to read cgroup item memory.kmem.limit_in_bytes
> Unable to read cgroup item blkio.throttle.io_serviced
> Unable to read cgroup item memory.kmem.usage_in_bytes
> Unable to read cgroup item memory.kmem.limit_in_bytes
> Unable to read cgroup item blkio.throttle.io_serviced
> Unable to read cgroup item memory.kmem.usage_in_bytes
> Unable to read cgroup item memory.kmem.limit_in_bytes
> Unable to read cgroup item blkio.throttle.io_serviced
> Unable to read cgroup item memory.usage_in_bytes
> Segmentation fault (core dumped)
> ```
> 
> I have a backtrace:
> 
> ```
> [New LWP 24554]
> Core was generated by `lxc-top'.
> Program terminated with signal SIGSEGV, Segmentation fault.
> #0  0x00007f3667931573 in strlen (s=<optimized out>, s@entry=0x0) at src/string/strlen.c:17
> 
> warning: 17     src/string/strlen.c: No such file or directory
> (gdb) 
> #0  0x00007f3667931573 in strlen (s=<optimized out>, s@entry=0x0) at src/string/strlen.c:17
> #1  0x000056275c5f4987 in stat_get_blk_stats (c=c@entry=0x7f366770fd60, item=item@entry=0x56275c5f9a27 "io.stat", stats=0x7f36676d1698, success=success@entry=0x7ffdbd023f1f)
>     at ../src/lxc/tools/lxc_top.c:307
> #2  0x000056275c5f706f in stats_get (total=0x7ffdbd023f40, ct=0x7f36676d1c30, c=0x7f366770fd60) at ../src/lxc/tools/lxc_top.c:361
> #3  lxc_top_main (argc=<optimized out>, argv=<optimized out>) at ../src/lxc/tools/lxc_top.c:663
> ```
> 
> Segfault happens due to `cols[1]` is `NULL` apparently.
> 
> ```c
>                if (strncmp(cols[1], "Read", strlen(cols[1])) == 0)
>                         stats->read += strtoull(cols[2], NULL, 0);
> ```

I will try on my end

DevonSchwartz commented:
I do not get this on my end. What steps did you take on your end to reproduce this?

ncopa commented:
I was not able to reproduce the segfault on a clean/fresh system. It happened on a relatively new server with containers imported from old system.

That said, `lxc-top` does not show any running containers either. Here is how to reproduce in qemu on x86_64:

Boot alpine in qemu:
```
qemu-system-x86_64 --enable-kvm -m 2048 -cdrom https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/x86_64/alpine-virt-3.20.2-x86_64.iso -nographic
```

Log in as root and run the following commands:
```
setup-alpine -q
apk add lxc lxc-dbg musl-dbg lxc-templates-legacy-alpine lxc-bridge iptables strace
rc-service cgroups start
rc-service dnsmasq.lxcbr0 start
lxc-create -t alpine -n a1
lxc-start -n a1
lxc-top -b
```

The output is:
```
alpine:~# lxc-top -b
time_ms,container,cpu_nanos,cpu_sys_userhz,cpu_user_userhz,blkio_bytes,blkio_iops,mem_used_bytes,mes
Unable to read cgroup item io.stat
Unable to read cgroup item blkio.throttle.io_service_bytes
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
```

To confirm that `a1` is running:
```
alpine:~# lxc-info -n a1
Name:           a1
State:          RUNNING
PID:            2596
IP:             10.0.3.86
Link:           vethGVKXS5
 TX bytes:      1.82 KiB
 RX bytes:      2.00 KiB
 Total bytes:   3.82 KiB
```

I would have expected that `lxc-top` would show that `a1` is running.

DevonSchwartz commented:
How do you run it on qemu using the change from this PR?

ncopa commented:
The patch was backported to the alpine 3.20 package, so it is included in `lxc-6.0.0-r3` and later: https://github.com/alpinelinux/aports/commit/1b2d32887a8dbb6a119f5c5d9078d60346bdffa7

DevonSchwartz commented:
Oh okay. I was testing it on alpine 3.17, so it could definitely be that my solution is not fit for later versions

DevonSchwartz commented:
I get the same thing you do on 3.20. For reference, in my forked repo I get this:

`time_ms,container,cpu_nanos,cpu_sys_userhz,cpu_user_userhz,blkio_bytes,blkio_iops,mem_used_bytes,memsw_used_bytes,kernel_mem_used_bytes
Unable to read cgroup item memory.kmem.usage_in_bytes
Unable to read cgroup item memory.kmem.limit_in_bytes
Unable to read cgroup item blkio.throttle.io_serviced
1721745840544,u2,0,5145694023161373743,8101259050997526833,0,0,5337088,0,0`

DevonSchwartz commented:
My guess is that the container stats are not being updated. If the paths were bad, then it would print "Unable to read cgroup item...". The container stats are set starting on line 371 (`if (total)... `). Could I use gdb to backtrace an lxc-top command?

DevonSchwartz commented:
I tried using `sudo ./lxc-create -t download a1` with alpine version 3.20, and I got the container statistics. However, when I use qemu (@ncopa 's setup), the statistics do not show up. Is it possible for there to be a discrepancy between an lxc template and running a released version of lxc on alpine?

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -276,27 +276,33 @@ static uint64_t stat_match_get_int(struct lxc_container *c, const char *item,
 	Total 149327872
 */
 static void stat_get_blk_stats(struct lxc_container *c, const char *item,
-			      struct blkio_stats *stats) {
+			      struct blkio_stats *stats, bool *success) {
```
Would it be an idea to do let the function return a `bool` for success?
```c
static bool stat_get_blk_stats(struct lxc_container *c, const char *item,
			       struct blkio_stats *stats) {
```

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -276,27 +276,33 @@ static uint64_t stat_match_get_int(struct lxc_container *c, const char *item,
 	len = c->get_cgroup_item(c, item, buf, sizeof(buf));
 	if (len <= 0 || (size_t)len >= sizeof(buf)) {
 		fprintf(stderr, "Unable to read cgroup item %s\n", item);
+		*success = false; 
 		return;
```
```suggestion
		return false;
```

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -314,21 +320,53 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
 	return;
 }
 
+static void try_cgroup2(struct lxc_container *c, u_int64_t *stat, const char* path_cgroup1, const char* path_cgroup2, 
+	const char* match, bool call_match) {
+	
+	int ret_cgroup2; 
+	
+	if (call_match) {
+		ret_cgroup2 = stat_match_get_int(c, path_cgroup2, match, 1); 
+		if (ret_cgroup2 < 0) {
+			*stat = stat_match_get_int(c, path_cgroup1, match, 1);
+		}
+		
+	} else {
+		ret_cgroup2 = stat_get_int(c, path_cgroup2); 
+		
+		if (ret_cgroup2 < 0) {
+			*stat = stat_get_int(c, path_cgroup1);
+		} else {
+			*stat = ret_cgroup2; 
+		}
+	}
+}
+
```
We can easily [avoid using boolean argument](https://google.com/search?q=avoid+boolean+arguments) here.

```suggestion
static void try_match_cgroup2(struct lxc_container *c, u_int64_t *stat, const char* path_cgroup1, const char* path_cgroup2, 
	const char* match) {
	
	int ret_cgroup2; 
	ret_cgroup2 = stat_match_get_int(c, path_cgroup2, match, 1); 
	if (ret_cgroup2 < 0) {
		*stat = stat_match_get_int(c, path_cgroup1, match, 1);
	}
}		

static void try_cgroup2(struct lxc_container *c, u_int64_t *stat, const char* path_cgroup1, const char* path_cgroup2) {
	ret_cgroup2 = stat_get_int(c, path_cgroup2); 
		
	if (ret_cgroup2 < 0) {
		*stat = stat_get_int(c, path_cgroup1);
	} else {
		*stat = ret_cgroup2; 
	}
}

```

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -314,21 +320,53 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
-
-	stat_get_blk_stats(c, "blkio.throttle.io_service_bytes", &ct->stats->io_service_bytes);
-	stat_get_blk_stats(c, "blkio.throttle.io_serviced", &ct->stats->io_serviced);
+	stat_get_blk_stats(c, "blkio.throttle.io_serviced", &ct->stats->io_serviced, &success);
```
```suggestion
	success = stat_get_blk_stats(c, "blkio.throttle.io_serviced", &ct->stats->io_serviced);
```

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -314,21 +320,53 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
+
+	// singular cgroup2 case for get blk stats
+	bool success; 
+	stat_get_blk_stats(c, "io.stat", &ct->stats->io_service_bytes, &success);
+	if (!success) {
+		stat_get_blk_stats(c, "blkio.throttle.io_service_bytes", &ct->stats->io_service_bytes, &success);
+	}
```
```suggestion
	success = stat_get_blk_stats(c, "io.stat", &ct->stats->io_service_bytes);
	if (!success) {
		success = stat_get_blk_stats(c, "blkio.throttle.io_service_bytes", &ct->stats->io_service_bytes);
	}
```

Could even be done as:
```c
	success = stat_get_blk_stats(c, "io.stat", &ct->stats->io_service_bytes) 
		|| stat_get_blk_stats(c, "blkio.throttle.io_service_bytes", &ct->stats->io_service_bytes);
```

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -314,21 +320,53 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
+	
+	if (call_match) {
+		ret_cgroup2 = stat_match_get_int(c, path_cgroup2, match, 1); 
+		if (ret_cgroup2 < 0) {
```
This wont work. `stat_match_get_int` returns a `uint64_t`, which is _unsigned_  int, meaning it will never be less than zero.

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -314,21 +320,53 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
+	} else {
+		ret_cgroup2 = stat_get_int(c, path_cgroup2); 
+		
+		if (ret_cgroup2 < 0) {
```
`uint64_t` will never be less than zero.

ncopa committed:
tools: lxc-top: refactor lxc-top stat structs

Create separate structs for each controller class. This will make it
easier to add cgroupv2 support.

No functional changes.

Signed-off-by: Natanael Copa <ncopa@alpinelinux.org>

ncopa committed:
tools: lxc-top: get memory stats from cgroups2

Signed-off-by: Natanael Copa <ncopa@alpinelinux.org>

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -314,21 +320,53 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
+	try_cgroup2(c, &(ct->stats->mem_limit), "memory.limit_in_bytes", "memory.max", NULL, false);
+	try_cgroup2(c, &(ct->stats->memsw_used), "memory.memsw.usage_in_bytes", "memory.swap.current", NULL, false);
+	try_cgroup2(c, &(ct->stats->memsw_limit), "memory.memsw.limit_in_bytes", "memory.swap.max", NULL, false);
+	try_cgroup2(c, &(ct->stats->cpu_use_nanos), "cpuacct.usage", "cpu.stat", NULL, false);
+	try_cgroup2(c, &(ct->stats->cpu_use_user), "cpuacct.stat", "cpu.stat", "user", true); 
+	try_cgroup2(c, &(ct->stats->cpu_use_sys), "cpuacct.stat", "cpu.stat", "system", true); 
```
for cgroups2 the match strings to look for in `cpu.stat` are:
- `usage_usec`
- `user_usec`
- `system_usec`

DevonSchwartz replied:
Thanks. I’ll put those and then send the result.

@ncopa are there any other match strings that I have incorrectly?

ncopa replied:
It's more complicated than that. the `io.stat` may look like this:
```
259:0 rbytes=0 wbytes=5094400000 rios=0 wios=4276 dbytes=0 dios=0
253:0 rbytes=0 wbytes=10491523072 rios=0 wios=1656 dbytes=0 dios=0
```
So you cannot simply look for the key as coloumn 1.

ncopa committed:
lxc-top: CPU stats for cgroups2

Recalculate the usec to nanoseconds and USER_HZ

Signed-off-by: Natanael Copa <ncopa@alpinelinux.org>

ncopa committed:
tools: lxc-top: get the user HZ at runtime

The USER_HZ depends on the kernel configuration. Get it run-time instead
of assume it is 100 HZ.

Signed-off-by: Natanael Copa <ncopa@alpinelinux.org>

ncopa committed:
tools: lxc-top: add cgroups2 IO stats

Signed-off-by: Natanael Copa <ncopa@alpinelinux.org>

ncopa commented:
Please also note that the units for CPU stats are different. In cgroups v1 it is nanoseconds and USER_HZ. In v2 it is usec.

And there is currently a bug. The `USER_HZ` is hardcoded to 100, but it depends on the running kernel, so it needs to be detected at run-time.

What do you think about something like this:
https://github.com/ncopa/lxc/commits/lxc-top-2/

DevonSchwartz commented:
No worries @ncopa. As a student, I am always appreciate of whatever help I can get. I am going to try your changes on my end and merge them. I'll also try to see if I can find a way to detect USER_HZ at runtime.

DevonSchwartz commented:
@ncopa, is it okay if I merge your branch into mine and fix the cpu stat issue?

DevonSchwartz commented:
I found this link in documentation: https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst

DevonSchwartz committed:
lxc-top: added kernel memory usage for cgroup2

Signed-off-by: Devon Schwartz <devon.s.schwartz@gmail.com>

ncopa commented:
> @ncopa, is it okay if I merge your branch into mine and fix the cpu stat issue?

Sure!

ncopa commented:
Good catch on the kernel mem fix! I missed that in the docs. The [lxc-info PR](https://github.com/lxc/lxc/pull/4426) calculates it differently and can probably be simplified.

You have the kernel docs into nicely rendered html as well: https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html

One comment on your commit message. It is usually a good idea to write the commit message in [imperative mood](https://google.com/search?q=git+commit+message+imperative).

ncopa commented on src/lxc/tools/lxc_top.c:
```diff
@@ -308,40 +314,130 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
+
+	if (cg1_get_blk_stats(c, "blkio.throttle.io_service_bytes", &ct->stats->io_service_bytes) < 0) {
+		if (cg2_get_blk_stats(c, "io.stat", &ct->stats->io_service_bytes) < 0) {
+			fprintf(stderr, "Unable to read CPU stats\n");
```
I did a copy/pasta error:
```suggestion
			fprintf(stderr, "Unable to read IO stats\n");
```

ncopa commented:
this also fixes https://github.com/lxc/lxc/issues/4337

ncopa commented:
I found another bug in `lxc-top` with `-b`. The `ct_print_cnt` should probably be ignored with `batch` mode.
```diff
diff --git a/src/lxc/tools/lxc_top.c b/src/lxc/tools/lxc_top.c
index 6cb38bb90..7498331ee 100644
--- a/src/lxc/tools/lxc_top.c
+++ b/src/lxc/tools/lxc_top.c
@@ -732,7 +732,7 @@ int lxc_top_main(int argc, char *argv[])
                        stats_print_header(&total);
                }
 
-               for (i = 0; i < active_cnt && i < ct_print_cnt; i++) {
+               for (i = 0; i < active_cnt && (i < ct_print_cnt || batch); i++) {
                        stats_print(container_stats[i].c->name, container_stats[i].stats, &total);
                        printf("\n");
                }
```

DevonSchwartz commented:
@ncopa have we figured out why container stats wont print on a alpine distro?

ncopa commented:
Yes, I believe it is due to: https://github.com/lxc/lxc/pull/4439#issuecomment-2250154083

ncopa commented:
Also, when running in qemu, diskless, there is no disk io, which gave "Unable to get CPU stats", where "CPU" was a copy/paste typo of "IO".

You can do: `git rebase -i main`, then set `e` or `edit`in front of the `tools: lxc-top: add cgroups2 IO stats`. Then you edit the file and fix the typo, and do `git add src/lxc/tools/lxc_top.c && git commit --amend -v`. Finally `git rebase --continue`

ncopa commented:
> @ncopa have we figured out why container stats wont print on a alpine 3.20 distro?

correction. No, I still don't know why it does not print on alpine 3.20 distro (without `-b`) . I suspect it is related the `ct_print_cnt`.

ncopa commented:
> > @ncopa have we figured out why container stats wont print on a alpine 3.20 distro?
> 
> correction. No, I still don't know why it does not print on alpine 3.20 distro (without `-b`) . I suspect it is related the `ct_print_cnt`.

I believe that the serial console does not give the window size of the terminal. 

But it also makes me think that we should probably get the window size in each iteration in the loop, because it is possible to resize the window while its running.

That is different issue though.

DevonSchwartz committed:
tools: lxc-top: add batch as conditional because ct_print_cnt irrelevant for lxc-top -b

Signed-off-by: Devon Schwartz <devon.s.schwartz@gmail.com>

DevonSchwartz commented:
I pushed the commits. Should we wait to squash commits until it is backported to alpine?

ncopa commented:
I think this is ready for merge. Personally I don't think the commits needs to be squashed, they give a good history in small steps. (and I have already backported this to alpine, even to stable branches)

@mihalicyn can you please review?

stgraber commented:
Looks like not everything is signed-off properly.

mihalicyn commented on src/lxc/tools/lxc_top.c:
```diff
@@ -631,7 +732,7 @@ int lxc_top_main(int argc, char *argv[])
 		}
 
-		for (i = 0; i < active_cnt && i < ct_print_cnt; i++) {
+		for (i = 0; i < active_cnt && i < (ct_print_cnt || batch); i++) {
```
Please, can you explain this change?

DevonSchwartz replied:
You're right. I tried on my end and it only prints the container statistics of one container. Thanks for catching that @mihalicyn 

```
Name:           a2
State:          RUNNING
PID:            4916
Link:           vethq4uXbP
 TX bytes:      9.20 KiB
 RX bytes:      0 bytes
 Total bytes:   9.20 KiB
```

```
time_ms,container,cpu_nanos,cpu_sys_userhz,cpu_user_userhz,blkio_bytes,blkio_iops,mem_used_bytes,memsw_used_bytes,kernel_mem_used_bytes
1722022106544,a1,280167000,17,10,2785280,0,4681728,0,1327104
``` 

In the above example, a2 is running, but only a1 prints

mihalicyn commented on src/lxc/tools/lxc_top.c:
```diff
@@ -35,16 +34,24 @@ struct blkio_stats {
+struct mem_stats {
+	uint64_t used;
+	uint64_t limit;
+	uint64_t swap_used;
```
it's not a swap, really. It's mem+swap. In cgroupv1 we have a combined page counter for memory+swap:
https://github.com/torvalds/linux/blob/1722389b0d863056d78287a120a1d6cadb8d4f7b/mm/memcontrol.c#L1536

for cgroup-v2, yes, we have a separate `swap` page counter.

I would go the same way, as it's done in the kernel:
https://github.com/torvalds/linux/blob/1722389b0d863056d78287a120a1d6cadb8d4f7b/include/linux/memcontrol.h#L191

and use union:
```
	union {
                struct {
                        uint64_t swap_used;
	                uint64_t swap_limit;
                }; /* v2 only */
                struct {
                        uint64_t memsw_used;
	                uint64_t memsw_limit;
                }; /* v1 only */
	};
```

it makes it a bit cleaner.

DevonSchwartz replied:
In cgroup2, should we keep the `memsw_used_bytes` column in `lxc-top -b` and calculate that value by adding `memory.current + memory.swap.current`?

mihalicyn commented on src/lxc/tools/lxc_top.c:
```diff
@@ -204,7 +212,6 @@ static uint64_t stat_get_int(struct lxc_container *c, const char *item)
 
 	len = c->get_cgroup_item(c, item, buf, sizeof(buf));
 	if (len <= 0) {
-		fprintf(stderr, "Unable to read cgroup item %s\n", item);
```
can we keep these error messages? I think it might be very-very useful in case if something goes wrong.

mihalicyn commented on src/lxc/tools/lxc_top.c:
```diff
@@ -308,40 +314,130 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
+	}
+
+	if (cg1_get_blk_stats(c, "blkio.throttle.io_service_bytes", &ct->stats->io_service_bytes) < 0) {
+		if (cg2_get_blk_stats(c, "io.stat", &ct->stats->io_service_bytes) < 0) {
```
ah, now I understand why you have removed these error messages :-)

I think you can detect if cgroup1/cgroup2 is used one time, save it in a variable and then call the right function instead of trying each time.

DevonSchwartz replied:
Is it possible that a system could be using cgorup1 and cgroup2 if the rc cgroup mode is set to "hybrid"?

mihalicyn commented:
@DevonSchwartz @ncopa Huge thanks for your work on this!

DevonSchwartz committed:
tools: lxc-top: print when batch or less than container count

Signed-off-by: Devon Schwartz <devon.s.schwartz@gmail.com>

DevonSchwartz commented on src/lxc/tools/lxc_top.c:
```diff
@@ -308,40 +314,130 @@ static void stat_get_blk_stats(struct lxc_container *c, const char *item,
+		}
+	} else {
+		/* only with cgroups v1 */
+		cg1_get_blk_stats(c, "blkio.throttle.io_serviced", &ct->stats->io_serviced);
```
Is it possible to calculate the fields in `ct->stats->io_serviced` by taking all the fields from `ct->stats->io_serviced_bytes` and dividing by `sizeof char`? This would be for cgroup2 since `io_serviced` is not a path in that cgroup

DevonSchwartz committed:
tools: lxc-top: add union to cgroup1 memsw and cgroup2 swap stats

Signed-off-by: Devon Schwartz <devon.s.schwartz@gmail.com>

paulocoghi commented:
Hello @DevonSchwartz! Is there any pending doubt you would like to ask? If yes, I would like to cordially ask you to tag the correspondent contributor related to your doubt, so we can have your PR merged :slightly_smiling_face:

DevonSchwartz commented:
There is no pending doubt.

paulocoghi commented:
@DevonSchwartz thanks for the fast response!

Also, I would like your opinion (also @ncopa's and @mihalicyn's opinion) on using this simple command to check which cgroup version is being used.

As per https://github.com/admiracloud/admira-containers/blob/develop/src/cli/requisites.cr#L5, by using:

```
stat -fc %T /sys/fs/cgroup/
```

You will get `cgroup2fs` when cgroups v2 is being used and `tmpfs` when v1 is being used.

Wouldn't this strategy be simpler than "trial and error" checks?

paulocoghi commented:
Thanks for the important clarification Stéphane @stgraber

mihalicyn reviewed (approved):
LGTM

Huge thanks to @DevonSchwartz and @ncopa!
